### PR TITLE
[Feat] Add support for tarball url as chart source

### PR DESCRIPTION
**Pull request by ghost**
**What is the purpose of this pull request?**: To support downloading/decompressing/extracting a tarball from a url and using as a chart source

**What issue does this pull request address?**: Fixes #

**Notes for reviewers to consider**:

**Specific reviewers for pull request**:
@gardlt @drewwalters96

### CODE REVIEW

**Review by intlabs — CHANGES_REQUESTED** [review]
@theyer could you provide an example yaml showing this in use.

**Comment by theyer** [comment]
@intlabs yes sorry this should actually be labeled as WIP; it was just created end of day Friday.

File: armada/utils/source.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    return extract_tarball(tarball_path)
+
+def download_tarball(tarball_url):
+    '''
+    Downloads a tarball to /tmp and returns the path
+    '''
+    try:
+        tarball_filename = path.basename(urlparse(tarball_url).path)
+        response = requests.get(tarball_url)
+        with open('/tmp/{}'.format(tarball_filename), 'wb') as f:
```
Comment (alanmeadows) [inline armada/utils/source.py]: Why note `tempfile` here like below?

Comment (theyer) [inline armada/utils/source.py]: my initial reasoning was to preserve the filename extension but after trying out tempfile, it seems it doesn't actually matter. I will switch this to use tempfile

File: armada/exceptions/source_exceptions.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base_exception
+
+class SourceException(base_exception.ArmadaBaseException):
```
Comment (drewwalters96) [inline armada/exceptions/source_exceptions.py]: Can you included these new exceptions in the exception troubleshooting documentation?

File: armada/utils/source.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+def download_tarball(tarball_url):
+    '''
+    Downloads a tarball to /tmp and returns the path
+    '''
+    try:
+        tarball_filename = tempfile.mkstemp(prefix='armada', dir='/tmp')[1]
+        response = requests.get(tarball_url)
+        with open(tarball_filename, 'wb') as f:
+            f.write(response.content)
+        return tarball_filename
```
Comment (drewwalters96) [inline armada/utils/source.py]: Should this be included in the try statement? If an exception is thrown, it is never reached regardless.

File: armada/utils/source.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    '''
+    if not path.exists(tarball_path):
+        raise source_exceptions.InvalidPathException(tarball_path)
+
+    _tmp_dir = tempfile.mkdtemp(prefix='armada', dir='/tmp')
+
+    try:
+        file = tarfile.open(tarball_path)
+        file.extractall(_tmp_dir)
+        return _tmp_dir
```
Comment (drewwalters96) [inline armada/utils/source.py]: Same as above: could this be moved outside of the try block?

**Review by drewwalters96 — CHANGES_REQUESTED** [review]
Looks good, small changes

File: armada/utils/source.py (hunk 3 above)
Comment (theyer) [inline armada/utils/source.py]: @drewwalters96 I was following suit with your exception handling changes which also included the return statement in the try block...

Comment (drewwalters96) [inline armada/utils/source.py]: In #111, I included return statements that could potentially throw an exception in try blocks. It seems here that the exception would be raised prior to the return statement.

Not a serious issue, just a nit.

File: armada/exceptions/source_exceptions.py (hunk 2 above)
Comment (theyer) [inline armada/exceptions/source_exceptions.py]: Done

**Review by drewwalters96 — APPROVED** [review]
LGTM

File: armada/utils/source.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        raise source_exceptions.GitLocationException(repo_url)
+
+    return _tmp_dir
+
+def get_tarball(tarball_url):
+    tarball_path = download_tarball(tarball_url)
+    return extract_tarball(tarball_path)
+
+def download_tarball(tarball_url):
+    '''
```
Comment (gardlt) [inline armada/utils/source.py]: does this mean that subpath for tar types are not used?

Comment (theyer) [inline armada/utils/source.py]: they can be. for example, in tar_example.yaml when the mariadb-0.1.0.tgz is extracted, the first directory level inside the tarball is 'mariadb' so the yaml has the subpath as 'mariadb'

Comment (theyer) [inline armada/utils/source.py]: if a tarball will just open right up into the content, the subpath should be omitted.

**Review by gardlt — APPROVED** [review]

**Pull request merged**